src/dotty/tools/dotc/core/Constraint.scala

```diff
@@ -10,451 +10,143 @@ import printing.Texts._
-      for (i <- 0 until paramCount(entries))
-        if (isBounds(entries(i)) && !p(PolyParam(poly, i))) return false
-    }
-    true
-  }
+  /** Check whether predicate holds for all parameters in constraint */
+  def forallParams(p: PolyParam => Boolean): Boolean
 
   /** Perform operation `op` on all typevars, or only on uninstantiated
    *  typevars, depending on whether `uninstOnly` is set or not.
    */
-  def foreachTypeVar(op: TypeVar => Unit): Unit =
-    myMap.foreachBinding { (poly, entries) =>
-      for (i <- 0 until paramCount(entries)) {
-        typeVar(entries, i) match {
-          case tv: TypeVar if !tv.inst.exists => op(tv)
-          case _ =>
-        }
-      }
-    }
-
-  private var myUninstVars: mutable.ArrayBuffer[TypeVar] = null
+  def foreachTypeVar(op: TypeVar => Unit): Unit
 
   /** The uninstantiated typevars of this constraint */
-  def uninstVars: collection.Seq[TypeVar] = {
-    if (myUninstVars == null) {
-      myUninstVars = new mutable.ArrayBuffer[TypeVar]
-      myMap.foreachBinding { (poly, entries) =>
-        for (i <- 0 until paramCount(entries)) {
-          typeVar(entries, i) match {
-            case tv: TypeVar if isBounds(entries(i)) => myUninstVars += tv
-            case _ =>
-          }
-        }
-      }
-    }
-    myUninstVars
-  }
-
-  def constrainedTypesText(printer: Printer): Text =
-    Text(domainPolys map (_.toText(printer)), ", ")
+  def uninstVars: collection.Seq[TypeVar]
 
-  def constraintText(indent: Int, printer: Printer): Text = {
-    val assocs =
-      for (param <- domainParams)
-      yield (" " * indent) ~ param.toText(printer) ~ at(param).toText(printer)
-    Text(assocs, "\n")
-  }
-
-  override def toText(printer: Printer): Text = {
-    val header: Text = "Constraint("
-    val uninstVarsText = " uninstVars = " ~
-      Text(uninstVars map (_.toText(printer)), ", ") ~ ";"
-    val constrainedText =
-      " constrained types = " ~ constrainedTypesText(printer) ~ ";"
-    val constraintsText =
-      " constraint = " ~ constraintText(3, printer) ~ ")"
-    Text.lines(List(header, uninstVarsText, constrainedText, constraintsText))
-  }
-}
-
-trait ConstraintRunInfo { self: RunInfo =>
-  private var maxSize = 0
-  private var maxConstraint: Constraint = _
-  def recordConstraintSize(c: Constraint, size: Int) =
-    if (size > maxSize) {
-      maxSize = size
-      maxConstraint = c
-    }
-  def printMaxConstraint()(implicit ctx: Context) =
-    if (maxSize > 0) typr.println(s"max constraint = ${maxConstraint.show}")
+ /** Check that no constrained parameter in `pt` contains itself as a bound */
+  def checkNonCyclic(pt: PolyType, entries: Array[Type])(implicit ctx: Context): Unit
+  
+  /** Check that no constrained parameter contains itself as a bound */
+  def checkNonCyclic()(implicit ctx: Context): Unit
+  
+ /** Check that no constrained parameter contains itself as a bound,
+   *  either directly or indirectly. This should be not a stricter criterion
```

Comment:
Better phrasing (imho): "This should not be a stricter criterion than checkNonCyclic because transitivity should always be eliminated"